system/lib/wasmfs/backends/node_backend.cpp

```diff
@@ -22,45 +22,95 @@ int _wasmfs_node_readdir(const char* path,
 // Write `mode` and return 0 or an error code.
 int _wasmfs_node_get_mode(const char* path, mode_t* mode);
 
+// Write `size` and return 0 or an error code.
+int _wasmfs_node_stat_size(const char* path, uint32_t* size);
+int _wasmfs_node_fstat_size(int fd, uint32_t* size);
+
 // Create a new file system entry and return 0 or an error code.
 int _wasmfs_node_insert_file(const char* path, mode_t mode);
 int _wasmfs_node_insert_directory(const char* path, mode_t mode);
 
-// Write `size` and return 0 or an error code.
-int _wasmfs_node_get_file_size(const char* path, uint32_t* size);
+// Unlink the given file and return 0 or an error code.
+int _wasmfs_node_unlink(const char* path);
+int _wasmfs_node_rmdir(const char* path);
+
+// Open the file and return the underlying file descriptor.
+int _wasmfs_node_open(const char* path, const char* mode);
+
+// Close the underlying file descriptor.
+int _wasmfs_node_close(int fd);
 
 // Read up to `size` bytes into `buf` from position `pos` in the file, writing
 // the number of bytes read to `nread`. Return 0 on success or an error code.
 int _wasmfs_node_read(
-  const char* path, void* buf, uint32_t len, uint32_t pos, uint32_t* nread);
+  int fd, void* buf, uint32_t len, uint32_t pos, uint32_t* nread);
 
 // Write up to `size` bytes from `buf` at position `pos` in the file, writing
 // the number of bytes written to `nread`. Return 0 on success or an error code.
-int _wasmfs_node_write(const char* path,
-                       const void* buf,
-                       uint32_t len,
-                       uint32_t pos,
-                       uint32_t* nwritten);
+int _wasmfs_node_write(
+  int fd, const void* buf, uint32_t len, uint32_t pos, uint32_t* nwritten);
 
 } // extern "C"
 
-class NodeFile : public DataFile {
+// The state of a file on the underlying Node file system.
+class NodeState {
+  // Map all separate WasmFS opens of a file to a single underlying fd.
+  size_t openCount = 0;
```

Comment:
Should this be atomic? Or does the node backend have some kind of assurance on running in a single thread like the old FS did?

Reply:
Core WasmFS only ever calls backend methods while the corresponding file lock is held, so backends generally don't need to do their own additional synchronization unless they share state between files.

Reply:
Ah, right, thanks!